source/adapters/cuda/memory.hpp

```diff
@@ -35,136 +163,16 @@ struct ur_mem_handle_t_ {
-
-      size_t getMapOffset() const noexcept { return MapOffset; }
-
-      /// Returns a pointer to data visible on the host that contains
-      /// the data on the device associated with this allocation.
-      /// The offset is used to index into the CUDA allocation.
-      void *mapToPtr(size_t Size, size_t Offset,
-                     ur_map_flags_t Flags) noexcept {
-        assert(MapPtr == nullptr);
-        MapSize = Size;
-        MapOffset = Offset;
-        MapFlags = Flags;
-        if (HostPtr) {
-          MapPtr = static_cast<char *>(HostPtr) + Offset;
-        } else {
-          // TODO: Allocate only what is needed based on the offset
-          MapPtr = static_cast<void *>(malloc(this->getSize()));
-        }
-        return MapPtr;
-      }
-
-      /// Detach the allocation from the host memory.
-      void unmap(void *) noexcept {
-        assert(MapPtr != nullptr);
-
-        if (MapPtr != HostPtr) {
-          free(MapPtr);
-        }
-        MapPtr = nullptr;
-        MapSize = 0;
-        MapOffset = 0;
-      }
-
-      ur_map_flags_t getMapFlags() const noexcept {
-        assert(MapPtr != nullptr);
-        return MapFlags;
-      }
-    } BufferMem;
-
-    // Handler data for surface object (i.e. Images)
-    struct SurfaceMem {
-      CUarray Array;
-      CUsurfObject SurfObj;
-      ur_mem_type_t ImageType;
-
-      CUarray getArray() const noexcept { return Array; }
-
-      CUsurfObject getSurface() const noexcept { return SurfObj; }
-
-      ur_mem_type_t getImageType() const noexcept { return ImageType; }
-    } SurfaceMem;
-
-    struct ImageMem {
-      CUarray Array;
-      void *Handle;
-      ur_mem_type_t ImageType;
-      ur_sampler_handle_t Sampler;
-
-      CUarray get_array() const noexcept { return Array; }
-
-      void *get_handle() const noexcept { return Handle; }
-
-      ur_mem_type_t get_image_type() const noexcept { return ImageType; }
-
-      ur_sampler_handle_t get_sampler() const noexcept { return Sampler; }
-    } ImageMem;
-  } Mem;
+  std::variant<BufferMem, SurfaceMem, ImageMem> Mem;
 
   /// Constructs the UR mem handler for a non-typed allocation ("buffer")
   ur_mem_handle_t_(ur_context_handle_t Context, ur_mem_handle_t Parent,
-                   ur_mem_flags_t MemFlags, MemImpl::BufferMem::AllocMode Mode,
+                   ur_mem_flags_t MemFlags, BufferMem::AllocMode Mode,
                    CUdeviceptr Ptr, void *HostPtr, size_t Size)
-      : Context{Context}, RefCount{1}, MemType{Type::Buffer}, MemFlags{
-                                                                  MemFlags} {
-    Mem.BufferMem.Ptr = Ptr;
-    Mem.BufferMem.Parent = Parent;
-    Mem.BufferMem.HostPtr = HostPtr;
-    Mem.BufferMem.Size = Size;
-    Mem.BufferMem.MapSize = 0;
-    Mem.BufferMem.MapOffset = 0;
-    Mem.BufferMem.MapPtr = nullptr;
-    Mem.BufferMem.MapFlags = UR_MAP_FLAG_WRITE;
-    Mem.BufferMem.MemAllocMode = Mode;
+      : Context{Context}, RefCount{1}, MemType{Type::Buffer},
```

Comment:
This change is only for the HIP adapter. The cuda changes are only there because this PR is dependent on https://github.com/oneapi-src/unified-runtime/pull/961 . I have made this a draft. Once the other PR merges I will ask for reviews